crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,425 @@
+///
+/// If a write fails, the commiter will save the batch it meant to write and try to write them
+/// again at the next opportunity, potentially adding more rows to the batch if more have arrived
+/// in the interim.
+///
+/// This task will shutdown if canceled via the `cancel` token, or if the channel it receives data
+/// on has been closed by the handler for some reason.
+fn committer<H: Handler + 'static>(
+    db: Db,
+    mut rx: mpsc::Receiver<Indexed<H>>,
+    config: CommitterConfig,
+    metrics: Arc<IndexerMetrics>,
+    cancel: CancellationToken,
+) -> JoinHandle<()> {
+    spawn_monitored_task!(async move {
+        // The `poll` interval controls the maximum time to wait between commits, regardless of the
+        // amount of data available.
+        let mut poll = interval(config.commit_interval);
+        let mut cool = interval(COOLDOWN_INTERVAL);
+
+        // We don't care about keeping a regular cadence -- these intervals are used to guarantee
+        // things are spaced at out relative to each other.
+        poll.set_missed_tick_behavior(MissedTickBehavior::Delay);
+        cool.set_missed_tick_behavior(MissedTickBehavior::Delay);
+
+        // Buffer to gather the next batch to write. This may be non-empty at the top of a tick of
+        // the commiter's loop if the previous attempt at a write failed. Attempt is incremented
+        // every time a batch write fails, and is reset when it succeeds.
+        let mut attempt = 0;
+        let mut batch = vec![];
+
+        // Data for checkpoints that haven't been written yet. Note that `pending_rows` includes
+        // rows in `batch`.
+        let mut pending: BTreeMap<u64, Vec<H::Value>> = BTreeMap::new();
+        let mut pending_rows = 0;
+
+        // TODO: Track the watermark (keep track of out-of-order commits, and update the watermark
+        // table when we accumulate a run of contiguous commits starting from the current watermark).
+
+        loop {
+            tokio::select! {
+                // Break ties in favour of operations that reduce the size of the buffer.
+                biased;
+
+                _ = cancel.cancelled() => {
+                    info!(pipline = H::NAME, "Shutdown received, stopping committer");
+                    break;
+                }
+
+                // Time to write out another batch of rows, if there are any.
+                _ = poll.tick(), if pending_rows > 0 => {
+                    let Ok(mut conn) = db.connect().await else {
+                        warn!(pipeline = H::NAME, "Failed to get connection for DB");
+                        cool.reset();
+                        continue;
+                    };
+
+                    let guard = metrics
+                        .committer_gather_latency
+                        .with_label_values(&[H::NAME])
+                        .start_timer();
+
+                    while batch.len() < H::CHUNK_SIZE {
```

Comment:
If I am reading this correctly, we are writing at most CHUNK_SIZE every commit_interval. This doesn't feel like it could push for the maximum throughput.

Reply:
That's not quite right -- I think it's described in the doc comment for this function, but the commit interval is the upperbound on time between updates, and the cooldown interval (20ms or so) is the lowerbound.

If you accumulate lots of changes (e.g. you are backfilling, or there is a lot of activity on chain) then each pipeline will push out a chunk, wait 20ms, push out another chunk, and so on, until it's expended.

If there is not much activity on chain, then you might wait up to the poll interval for enough data to turn up, and then push it out, even if it isn't usually considered "enough".

Reply:
So if the previous write was successful, and there are still more than enough, why not commit right away instead of cooldown for 20ms?

Reply:
The intention was so that a single pipeline didn't hog the connection pool, i.e. we leave a small gap between when the connection is handed back and when we ask for a connection back, but we can experiment with reducing the delay or removing it.

Reply:
I added a comment to make sure we explore the possibility of removing this detail.